### ISSUE TITLE
UNDO is focus sensitive instead of global (like every other app)

### ISSUE THREAD

**Issue by jeske** [stride3d/stride#681]
**Release Type**: official 4.0.0.1 beta 2 0926
**Version**: official 4.0.0.1 beta 2 0926
**Platform(s)**: Windows

**Describe the bug**

Xenko UNDO is context-sensitive, which leads to non-intuitive behavior when trying to use the built-in editor... If focus is accidentally not in the editor, pressing undo can remove the entire file if the last action was to add the script, which is non-intuitive and dangerous. 

**To Reproduce**

1) add a script
2) edit the contents of the script (and dont save)
3) click in the solution explorer
4) press undo (ctrl-Z)
5) WITNESS: it removes the script from the project, blowing away edits

**Expected behavior**

EXPECTED: undo to undo last action globally (aka reverse edits) regardless of context (like almost every other application) 

EXPECTED: Xenko to ask before blowing away unsaved changes in an editor.

**Additional context**

Pretty much other editor environment I use (such as visual studio) has a single global UNDO stack.. which means if you press undo, it will remove the last action you did globally in the application, regardless of your focus context. It's very very non-intuitive that in Xenko, if you change the focus and undo, it will undo different things. 

If for some reason this multiple undo stacks and focus-aware undo is an essential feature, then at least it should warn me before removing a script, or doing some other massive undo operation.

**Comment by Kryptos-FR** [comment]
Incorrect. The undo/redo stack is global. The only exception is the built-in text editor that you used for code which has its own undo/redo capability that is separated from the global editor one.

**Comment by jeske** [comment]
If the editor has it's own undo stack, then the undo stack is not global. By definition. Since if it was global, there would be only one.

I think 90% of this problem could be avoided by just making a warning dialog for undo that will remove an asset that could result in loss of data (like an unsaved script file, because when it disappears one can't save it anymore). As long as the file is still on disk, and is fully saved, then one wouldn't really lose anything. 

I see your point about redo.. but in that moment where one is expecting it to undo a keystroke and Xenko eats the whole file and throws it in the trash.. it's more a moment of panic and there is no thought of redo. The thought is "omg why am i using software that will eat my data"

**Comment by Kryptos-FR** [comment]
I think our current solution is consistent with what any other application is doing. It would be unexpected to have a different behavior. And we don't want to have popups all over the place.

As far as I am concerned, nothing need to be changed, but I'll let others give their opinion on that matter.

**Comment by sebllll** [comment]
i have also seen my scriptfiles going to nirvana in some cases. a popup before deleting something permanently wouldn't be too much of a clutter in my opinion.

**Comment by jeske** [comment]
Neither Unity nor Visual Studio have an undo for file/asset additions as far as I can tell.  Adding a code-file or form in visual studio and pressing undo does nothing. Adding a script or prefab in unity and pressing undo does nothing. 

Maybe that is the simpler path to fix this. Just remove script additions from the undo stack entirely.

@Kryptos-FR Can you list the "other application" where you can "add a script, type code into script, change focus, press undo, and the script disappears"?
